### PR TITLE
Add a file provider that avoids ts program walk

### PR DESCRIPTION
**Pull request by Adjective-Object**
Adds a faster file provider that discovers source and fence files with a
fast asyncronous file walk rather than by relying on typescript to
perform a syncronous walk + parse of the program.

We still use typescript to parse compilerOptions, but since we no longer
rely on the typescript program walk to identify files, we stub out file
discovery so only the initial config parse happens.

Also adds config options to support switching between the two providers,
since it will only work when all source files you intend to check fences
against fall under your rootDirs.
This is the case for OWA, but I don't know what other consumers look
like.

Depends on u/mahuangh/faster-source-providers-2

Quick (completely nonscientific) test for impact on performance:
```
$ time yarn good-fences --rootDir packages shared -p ./tsconfig.noprojects.json --progressBar -x
Done in 45.17s.

real    0m45.452s
user    0m0.000s
sys     0m0.152s

$ time yarn good-fences --rootDir packages shared -p ./tsconfig.noprojects.json --progressBar
Done in 359.96s.

real    6m0.237s
user    0m0.046s
sys     0m0.121s
```

### CODE REVIEW

File: src/core/FdirSourceFileProvider.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    async getSourceFiles(searchRoots?: string[]): Promise<string[]> {
+        const allRootsDiscoveredFiles: string[][] = await Promise.all(
+            (searchRoots || this.rootDirs).map(
+                (rootDir: string) =>
+                    new fdir()
+                        .glob(
+                            this.parsedCommandLine.options.allowJs
+                                ? `**/!(.d)*@(.js|.ts${
```
Comment (Adjective-Object) [inline src/core/FdirSourceFileProvider.ts:66]: why do we exclude .d.ts here?

Comment (Adjective-Object) [reply src/core/FdirSourceFileProvider.ts:66]: Struggling to remember

File: src/utils/getOptions.ts [hunk 2]
```diff
@@ -27,7 +27,8 @@ export function setOptions(rawOptions: RawOptions) {
         project,
         rootDir,
         ignoreExternalFences: rawOptions.ignoreExternalFences,
+        looseRootFileDiscovery: rawOptions.looseRootFileDiscovery || false,
         maxConcurrentFenceJobs: rawOptions.maxConcurrentJobs || 6000,
-        progress: rawOptions.progressBar || false,
+        progress: rawOptions.progress || false,
```
Comment (Adjective-Object) [inline src/utils/getOptions.ts]: progressBar

File: src/core/FdirSourceFileProvider.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    async getSourceFiles(searchRoots?: string[]): Promise<string[]> {
+        const allRootsDiscoveredFiles: string[][] = await Promise.all(
+            (searchRoots || this.rootDirs).map(
+                (rootDir: string) =>
+                    new fdir()
+                        .glob(
+                            this.parsedCommandLine.options.allowJs
+                                ? `**/*@(.js|.ts${
```
Comment (smikula) [inline src/core/FdirSourceFileProvider.ts:59]: Nit: for clarity, maybe you can precalculate the extensions to glob for and `Array.join` them into the glob expression here?  (Also has the nominal benefit of not recalculating it for every rootDir.)

File: package.json [hunk 4]
```diff
@@ -17,7 +17,9 @@
     "dependencies": {
         "cli-progress": "^3.9.0",
         "commander": "^7.2.0",
+        "fdir": "^5.1.0",
```
Comment (smikula) [inline package.json:20]: >fdir [](http://example.com/codeflow?start=9&length=4)

Looks like we should add a dep on `picomatch`:
[https://github.com/thecodrr/fdir/blob/master/documentation.md#globstring](https://github.com/thecodrr/fdir/blob/master/documentation.md#globstring)

File: src/core/FdirSourceFileProvider.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            ) {
+                // this is an allowed script file
+                return directImportResult;
+            } else {
+                // this is an asset file
+                return undefined;
+            }
+        } else {
+            // resolve with tsconfig-paths (use the paths map, then fall back to node-modules)
+            return await new Promise((res, rej) =>
```
Comment (smikula) [inline src/core/FdirSourceFileProvider.ts:108]: >res, rej [](http://example.com/codeflow?start=38&length=8)

Nit: can you spell out the full words?

**Review by smikula — CHANGES_REQUESTED** [review]
This looks great.  I wish we didn't have to bend over backwards to get good perf relative to typescript, but the speed gains are impressive!  I just left a handful of nitpick comments.

File: package.json (hunk 4 above)
Comment (Adjective-Object) [reply package.json:20]: Good catch. I think I had glob installed alongside this prior which is why it was working locally.

File: src/core/FdirSourceFileProvider.ts (hunk 3 above)
Comment (Adjective-Object) [reply src/core/FdirSourceFileProvider.ts:59]: I think for consistency here I should update getScriptFileExtensions to do this calculation and call it here & below.

Comment (Adjective-Object) [reply src/core/FdirSourceFileProvider.ts:59]: I also end up needing a list of file extensions in #101, so I think deriving them from the same place is probably for the best.

**Review by smikula — APPROVED** [review]

**Pull request merged**